src/launchCommand.ts

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+import * as cp from "child_process";
+import * as vscode from "vscode";
+
+import { inferLaunchCommandLength } from "./languageServerPlugin";
+import { getJavaHome } from "./utility";
+
+enum shortenApproach {
+    none = "none",
+    jarmanifest = "jarmanifest",
+    argfile = "argfile",
+}
+
+export async function populateShortenCommandLineField(config: vscode.DebugConfiguration): Promise<vscode.DebugConfiguration> {
+    if (config.request === "launch" && (!config.shortenCommandLine || config.shortenCommandLine === "auto")) {
+        config.shortenCommandLine = await detectLaunchCommandStyle(config);
+    }
+
+    return config;
+}
+
+export async function detectLaunchCommandStyle(config: vscode.DebugConfiguration): Promise<shortenApproach> {
+    const cliLength = await inferLaunchCommandLength(config);
+    const javaHome = await getJavaHome();
+    const javaVersion = await checkJavaVersion(javaHome);
+    const recommendedApproach = javaVersion <= 8 ? shortenApproach.jarmanifest : shortenApproach.argfile;
+    if (config.console === "internalConsole") {
+        // In windows, the max process commmand line length is 32k (32678) characters.
+        // TODO Differentiate the limit for other platforms OSX and linux.
+        return cliLength < 32678 ? shortenApproach.none : recommendedApproach;
+    } else {
+        // In windows, the max command line length for cmd terminal is 8191 characters.
+        // TODO Differentiate the limit for other platforms OSX and linux.
+        return cliLength < 8191 ? shortenApproach.none : recommendedApproach;
+    }
+}
+
+function checkJavaVersion(javaHome: string): Promise<number> {

```

Comment:
I think this is the third place that we need parse/check java version. Please work with @Eskibear  & @jdneo for how we can consolidate this part logic together or put in the vscode java API part.

Reply:
agree. but prefer to do it with a new PR when the change in vscode-java is ready.